src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/StorePal.AppleKeychainStore.cs

```diff
@@ -68,13 +68,34 @@ public void Add(ICertificatePal cert)
 
             public void Remove(ICertificatePal cert)
             {
-                if (_readonly)
-                    throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
-
+                // First check if the certificate is even in the keychain
                 AppleCertificatePal applePal = (AppleCertificatePal)cert;
-
-                var handle = (SafeKeychainItemHandle)applePal.IdentityHandle ?? applePal.CertificateHandle;
-                Interop.AppleCrypto.X509StoreRemoveCertificate(handle, _keychainHandle);
+                using (SafeCFArrayHandle certs = Interop.AppleCrypto.KeychainEnumerateCerts(_keychainHandle))
+                {
+                    long count = Interop.CoreFoundation.CFArrayGetCount(certs);
+
+                    for (int i = 0; i < count; i++)
+                    {
+                        IntPtr itemHandle = Interop.CoreFoundation.CFArrayGetValueAtIndex(certs, i);
+
+                        SafeSecCertificateHandle certHandle;
+                        SafeSecIdentityHandle identityHandle;
+
+                        if (Interop.AppleCrypto.X509DemuxAndRetainHandle(itemHandle, out certHandle, out identityHandle))
+                        {
+                            if (applePal.IdentityHandle == identityHandle && applePal.CertificateHandle == certHandle)
+                            {
+                                if (_readonly)
+                                {
+                                    throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
+                                }
+
+                                var handle = (SafeKeychainItemHandle)applePal.IdentityHandle ?? applePal.CertificateHandle;
+                                Interop.AppleCrypto.X509StoreRemoveCertificate(handle, _keychainHandle);
+                            }
+                        }
+                    }
+                }
```

Comment:
I am not sure this is the best way to go about it because of the performance implications.

First, this penalizes the performance of the `!_readonly` case that was already handled correctly and without a loop over existing certificates. Second, the performance of the `_readonly` case could be made much better by using the whole `X509StoreRemoveCertificate`-like flow up to the PAL layer. In the PAL layer there's already a code for fast query of whether certificate is in store or not.

I'd suggest introducing `X509StoreLookupCertificate` (or different name, if you prefer). Wrapping it in the `Interop` classes in similar way and implementing in the PAL by reusing the code from `DeleteInKeychain` that creates the lookup query and just replacing the `SecItemDelete` call with `SecItemCopyMatching` for the read-only case.

Reply:
I wasn't particularly happy about this solution either, but I couldn't find a way to determine at the PAL layer to check for the presence of a particular certificate. I will have another look at this based on your information.

Reply:
It is definitely possible to check the presence by using something like:
```
    CFDictionarySetValue(query, kSecReturnRef, kCFBooleanTrue);
    CFDictionarySetValue(query, kSecMatchSearchList, searchList);
    CFDictionarySetValue(query, kSecMatchItemList, itemMatch);
    CFDictionarySetValue(query, kSecClass, kSecClassCertificate);
    *pOSStatus = SecItemCopyMatching(query, &result);
```

Kind of a mix between `EnumerateKeychain` and `DeleteInKeychain` in `pal_keychain.c`. I am not sure what is the best way to design it at the API level. @bartonjs may have some thought on that. The two options I considered are a adding a separate `X509StoreRemoveCertificate`-like API all the way through to the Interop and PAL level, or adding some `dryRun` parameter to `X509StoreRemoveCertificate` itself.

Reply:
I'd probably add a parameter to AppleCryptoNative_X509StoreRemoveCertificate (uint8_t isReadOnlyMode).

```C
if (!IsCertInKeychain(cert, keychain))
{
    return 1;
} 

if (isReadOnlyMode)
{
    // define this as const 4.
    return kErrorReadonlyDelete;
}
```

to https://github.com/dotnet/corefx/blob/225be960445683a97e68005d87f7ebbd5f5e5a88/src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_keychain.c#L351-L353.  Then handle that in the managed caller, because probably we also shouldn't return an admin/user trust modification problem if the cert isn't in the keychain that it's being asked to be removed from.

A more efficient model would be to only do the existence test when we need to.  So when in a read-only mode or when either user or admin trust rules were found.

So, perhaps, instead:

```C
int32_t ret = 0;

if (isReadOnlyMode)
{
    ret = kErrorReadonlyDelete;
}
else
{
    // Move current detection to a distinct function
    ret = CheckTrustSettings(cert);
}

if (ret != 0)
{
    if (!IsCertInKeychain(cert, keychain))
    {
        return 0;
    }

    return ret;
}

// current footer, for alignment
*pOSStatus = DeleteInKeychain(cert, keychain);
CFRelease(cert);
return *pOSStatus == noErr;
```

Reply:
I like the last code in the comment the most, but I think it still has unnecessary overhead for the non-read-only code path. You don't need to call both `IsCertInKeychain` and `DeleteInKeychain` at the same time since they would do exactly the same lookup.

Reply:
@filipnavara That's why the second version only checks when it was about to report an error (and was emphatically going to not call DeleteInKeychain).

Reply:
Ah, right!